File: Harvard_CS50_Introduction_to_CS/CS50-Pset6-master/crack.py
from sys import argv
from sys import exit
import crypt


def check(password, salt, h):
    encrypted = crypt.crypt(password, str(salt))
    if h == encrypted:
        print(password)
        exit(2)


# check the program running with one command-line argument
if len(argv) != 2:
    print("Usage: python crack.py hash")
    exit(1)
# obtain hash tring
h = argv[1]
salt = int(h[0] + h[1])
list_ascii = list(range(65, 91)) + list(range(97, 123))
guess = list()
# all password's characters can be combined with maximum number of chars - 5 as the below
# unfortunately, it will take so long time to crack a password with only 1 char
# since the nested loop will start from the lowest level of the loop
# it is easy to crack a password with 5 characters rather than less characters
# more over password starts with high index of ascii, the cracking talks even longer
for c1 in list_ascii:
    # crack the password having 1 char
    check(chr(c1), salt, h)
    for c2 in list_ascii:
        # crack the password having 2 chars
        temp2 = chr(c1) + chr(c2)
        check(temp2, salt, h)
        for c3 in list_ascii:
            # crack the password having 3 chars
            temp3 = chr(c1) + chr(c2) + chr(c3)
            check(temp3, salt, h)
            for c4 in list_ascii:
                # crack the password having 4 chars
                temp4 = chr(c1) + chr(c2) + chr(c3) + chr(c4)
                check(temp4, salt, h)
                # crack the password having 5 chars
                for c5 in list_ascii:
                    temp5 = chr(c1) + chr(c2) + chr(c3) + chr(c4) + chr(c5)
                    check(temp5, salt, h)

File: Harvard_CS50_Introduction_to_CS/CS50-Pset2-master/crack.c
#include <cs50.h>
#include <stdio.h>
#include <crypt.h>
#include <string.h>
#include <ctype.h>

int main(int argc, string argv[])
{
    //check if one command-line argument is executed
    if (argc != 2)
    {
        printf("Usage: ./crack hash");
        return 1;
    }
    char check(char c, int key);
    string hash = argv[1];
    char str_salt[3];
    for (int i = 0; i < 2; i++)
    {
        str_salt[i] = hash[i];
    }
    str_salt[2] = '\0';
    int salt = atoi(str_salt);

    //get ciphertext
    printf("ciphertext: ");
    int length = strlen(hash);
    for (int i = 2; i < length; i++)
    {
        printf("%c",check(hash[i],salt));
    }

    printf("\n");
    return 0;
}
//check lower or upper case chars
char check(char c, int key)
{
    if (isupper(c))
    {
        return ((c - 65 + key)%26 + 65);
    }
    else if (islower(c))
    {
        return ((c - 97 + key)%26 + 97);
    }
    else
    {
        return c;
    }
}